src/txn/txn_timestamp.c

```diff
@@ -109,30 +108,21 @@ __wt_txn_get_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *tsp, uin
     if (LF_ISSET(WT_TXN_TS_INCLUDE_CKPT) && txn_global->checkpoint_timestamp != WT_TS_NONE &&
       (tmp_ts == 0 || txn_global->checkpoint_timestamp < tmp_ts))
         tmp_ts = txn_global->checkpoint_timestamp;
-    if (!txn_has_write_lock)
-        __wt_readunlock(session, &txn_global->rwlock);
 
-    /* Look for the oldest ordinary reader. */
-    __wt_readlock(session, &txn_global->read_timestamp_rwlock);
-    TAILQ_FOREACH (txn_shared, &txn_global->read_timestamph, read_timestampq) {
-        /*
-         * Skip any transactions on the queue that are not active. Copy out value of read timestamp
-         * to prevent possible race where a transaction resets its read timestamp while we traverse
-         * the queue.
-         */
-        if (!__txn_get_read_timestamp(txn_shared, &tmp_read_ts))
-            continue;
+    /* Walk the array of concurrent transactions. */
+    WT_ORDERED_READ(session_cnt, conn->session_cnt);
+    WT_STAT_CONN_INCR(session, txn_walk_sessions);
+    for (i = 0, s = txn_global->txn_shared_list; i < session_cnt; i++, s++) {
+        __txn_get_read_timestamp(s, &tmp_read_ts);
         /*
          * A zero timestamp is possible here only when the oldest timestamp is not accounted for.
          */
-        if (tmp_ts == 0 || tmp_read_ts < tmp_ts)
+        if (tmp_ts == 0 || (tmp_read_ts != WT_TS_NONE && tmp_read_ts < tmp_ts))
```

Comment:
Not your change, but this should be using `WT_TS_NONE`. Same as the assignment at the beginning of this function.

Reply:
Fixed.